Introduce `Prologue` and `Discoveries` of an address book

HeinrichApfelmus opened the pull request:
### Issue number

ADP-1043

### Overview

Previous work in epic ADP-1043 introduced delta encodings, DBVars, and an embedding of the wallet state and its delta encodings into a database table. It's time to integrate these tools with the wallet code. To facilitate code review, the integration proceeds in a sequence of refactorings that do not change functionality and pass all unit tests.

In this step, we begin refactoring the address discovery state (the `s` in `Wallet s`). First, I would like to update our nomenclature and call this state an *address book*. Second, we introduce two new data families `Prologue s` and `Discoveries s`, with the intention that any address book is isomorphic to a pair of these two types, that is `s ~ (Prologue s, Discoveries s)`. The idea is that the Prologue of the address book contains "static" information such as public keys or the address gap (default 20), while the Discoveries of the address book contains information that was gathered during on-chain address discovery.

### Details

* We introduce a type class `AddressBookIso` which provides the isomorphism `s ~ (Prologue s, Discoveries s)`.
* For this pull request, we avoid making any changes to the `AddressDiscovery.*` hierarchy. Instead, the isomorphisms are implemented using newly defined data families whose representation is close to the types in the database.
* The implementation for the sequential states uses `liftPaymentAddress` to convert the payment part to and from the `Address` type. This is ok when convering to/from the database layer, but does not make sense for in-memory data structures. Therefore, further work in this epic will introduce changes in the `AddressDiscovery.*` hierarchy.

### Comments

* The overall goal is that, in a future PR, we can represent the wallet state along the lines of `WalletState s ~ (Prologue s, Checkpoints (BlockHeader, UTxO, Discoveries s) `.
* Merge PR #3048 before this one, because this pull request is based on the branch of the former.

Commit:
Introduce `Prologue` and `Discoveries` of an address state

In order to avoid making changes to the `AddressDiscovery.*` modules for the time being, we decompose the address state by means of two data families and an isomorphism.

* The `Prologue` captures off-chain information such as public keys or address gaps. This information is not associated with chain checkpoints.
* The `Discoveries` capture on-chain information such as discovered addresses in the `AddressPool`. This information has to be associated with chain checkpoints.

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite/AddressBook.hs:
```diff
@@ -0,0 +1,228 @@
+
+    -- | Isomorphism between the address book type 's'
+    -- and its two components.
+    addressIso :: Iso' s (Prologue s, Discoveries s)
```
:+1:

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite/AddressBook.hs:
```diff
@@ -0,0 +1,228 @@
+    Sequential address book
+-------------------------------------------------------------------------------}
+-- | Sequential list of addresses
+newtype SeqAddressList (c :: Role) = SeqAddressList [(W.Address, W.AddressState)]
```
Why we have here list? Because we are implying that the sequence of addresses correspond to address indices?So we have first entry corresponding to soft index ix=0, etc.? Why not to have a map with soft index as key? As we plan to enlarge this list (via `(++) [new address]`) or rollback usually with operations in the end of it...it seems the list could be quite expensive - naively thinking. Maybe strict map would be better here and we would not loose the capabilities here? What do you think about this argument?

HeinrichApfelmus replied:
I completely agree that we will need a more efficient data structure than lists here — which is why I started refactoring the address pool logic, and was able to implement it for `SharedStates` in the upcoming PR #3068 😄. Step by step.

For the present PR, the list is sufficient: At the moment, the list is created when loading from the database and then immediately converted into an address pool through the `addressIso`. Similar procedure when storing to the database. As long as the conversion with `addressIso` happens just at these points, immediately before or after the database is accessed, we do pretty much exactly the same thing that the previous version of our code did, so we do not pay a performance penalty for now. Of course, once the isomorphism is used in the context of `Checkpoints (Discoveries s, UTxO,…)`, we need it to be faster. But that is for another PR.

paweljakubas replied:
ok. fair enough

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite/AddressBook.hs:
```diff
@@ -0,0 +1,228 @@
+    , Typeable c
+    ) => Seq.AddressPool c key -> SeqAddressList c
+toDiscoveries pool = SeqAddressList
+    [ (a,st) | (a,st,_) <- Seq.addresses (liftPaymentAddress @n) pool ]
```
We do like this because SeqAddressList is directly mapped to DB, right?

paweljakubas replied:
if yes, then for the future reference we need to remember about some probable future developments here:
1. multi accountsupport - in that case liftPaymentAddress will not be enough as payment credential is not enough. We would need to also differentiate against staking credential so staking key hash (for Sequential mode) or staking script hash (for Shared mode)
2. we have (at least in Sequential mode) several fingerprints of address to support: the current one (aka base with key hash in payment credential), enterprise address (key hash payment credenital without staking credential) and pointer address (key hash payment with pointer stake credential)
We need (probably later) to discuss if new abstraction are able to seamlessly support them

HeinrichApfelmus replied:
> We do like this because SeqAddressList is directly mapped to DB, right?

Yes, indeed. Will be changed again in the next steps.

> […] staking credential […]

That's a very good point. From looking at the code, I got the impression that the Sequential state only looks at the payment key hashes. In other words, whenever the state sees an address on chain (which may have a stake key hash), it will extract the payment key hash, add it to the address pool as appropriate, convert it to an enterprise address and store that address in the database. Put differently, the database stores discovered Sequential payment key hashes, but formatted as enterprise addresses.

The situation for Shared states was not entirely clear to me, but by chasing the various type classes, I also got the impression that only the payment key hash is subject to discovery. In other words, if an address appears on the chain, and the stake credential of this address does not match the delegation template, but the payment credential does match the payment template, then it will still be discovered and added to the database. Makes sense: If you own the payment credential, you can always reassign the delegation part by making a transaction; the access control on the payment credential partially overrides the access control on the delegation part. I suppose we will need to think about this corner case again. 🤔 

In any case, for the time being, this PR only attempts to do what the code did before, no changes in functionality intended. 😅

paweljakubas replied:
so for shared wallets we store payment credential only which is script hash. shared wallet's credential and sequential wallet credential happen to have the same byte imprint as the same hashes (blake224) in both cases is used.

What is important here is that maybe to avoid complications and choose simplicity it would be better to store both credentials, whatever they are represented (keyhash, scripthash or three numbers for pointer addresses). It would allow support of any address type and also seamlessly support multiaccount (on the level of discovery) 

I feel we will need to revisit this soon or later (maybe even during shared wallet work soon). I also see challenges (db migration issues....)

HeinrichApfelmus replied:
> I feel we will need to revisit this soon or later (maybe even during shared wallet work soon). I also see challenges (db migration issues....)

Agreed. I suggest that we leave things as they are for now and revisit this soon during the shared wallet work.

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite/AddressBook.hs:
```diff
@@ -0,0 +1,228 @@
+    , key ~ SharedKey
+    ) => Seq.AddressPool c key -> SeqAddressList c
+toDiscoveriesShared pool = SeqAddressList
+    [ (a,st) | (a,st,_) <- Seq.addresses (Shared.liftPaymentAddress @n) pool ]
```
the same comment as before... if we want to cover all possible addresses having script hashes in credentials this could not be 1-to-1 mapping anymore also here

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite/AddressBook.hs:
```diff
@@ -0,0 +1,228 @@
+            Shared.PendingFields{} -> st
+            Shared.ReadyFields pool0
+                -> Shared.SharedState a $ Shared.ReadyFields
+                $ fromDiscoveries @n pool0 addrs
```
ok, should be fine :+1:

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite/AddressBook.hs:
```diff
@@ -0,0 +1,228 @@
+
+-- | Isomorphism for HD random address book.
+instance AddressBookIso (Rnd.RndState n) where
+    data Prologue (Rnd.RndState n)
```
will look at this during the second look tomorrow. Looks correct at first sight

paweljakubas replied:
looks fine to my best knwoledge

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite/CheckpointsOld.hs:
```diff
@@ -225,19 +234,20 @@ mkStoreCheckpoints wid =
         [ CheckpointWalletId ==. wid ]
         [ Desc CheckpointSlot ]
-    fmap catMaybes $ forM cps $ \cp -> do
+    -- FIXME LATER during ADP-1043: Presence of these tables?
```
would be good to have check on if the tables exist and only after they are proceed :+1:

HeinrichApfelmus replied:
Good point. I have added a commit where a new error type `ErrBadFormat` is introduced, which I intend to eventually cover all errors related to malformed database files. As a first step, it absorbs the `fromJust` case here.

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite/CheckpointsOld.hs:
```diff
@@ -225,19 +234,20 @@ mkStoreCheckpoints wid =
         [ Desc CheckpointSlot ]
-    fmap catMaybes $ forM cps $ \cp -> do
+    -- FIXME LATER during ADP-1043: Presence of these tables?
+    prologue <- fromJust <$> loadPrologue wid
```
so here we have assumption that static part (aka Prologue) never changes ... we will need to be careful here when adding support for multi accounts... list of acctXPub, so a given accXPub, could stop to be static...just writing for future ourselves;-)

HeinrichApfelmus replied:
Yes. Perhaps a good way to think about is that the "static part" is only "static" in the sense that it does not depend on information that discovered on the blockchain, but not "static" in the sense that it can never change. That's one of the reasons why I have called it "prologue" instead of "static". 😄

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite/CheckpointsOld.hs:
```diff
@@ -225,19 +234,20 @@ mkStoreCheckpoints wid =
-            <$> st
+        discoveries <- loadDiscoveries wid (checkpointSlot cp)
+        let st = withIso addressIso $ \_ from -> from (prologue, discoveries)
+        pure $ let c = checkpointFromEntity @s cp utxo st in (getPoint c, c)
```
would be in favor of extracting let in the above line. just for the sake of clarity

HeinrichApfelmus replied:
Added in the additional commit.

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite/CheckpointsOld.hs:
```diff
@@ -349,44 +362,45 @@ checkpointFromEntity cp (coins, tokens) =
-    -- | Load the state for a checkpoint.
-    selectState :: (W.WalletId, W.SlotNo) -> SqlPersistT IO (Maybe s)
+-- | Functions for saving / loading the wallet's address book to / from SQLite
+class AddressBookIso s => PersistAddressBook s where
```
I am fond of this abstraction :100:

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite/CheckpointsOld.hs:
```diff
@@ -475,21 +464,33 @@ selectSeqStatePendingIxs wid =
 
+selectSeqAddressList
+    :: forall c. Typeable c
+    => W.WalletId -> W.SlotNo -> SqlPersistT IO (SeqAddressList c)
```
ok, so I have answer here about relation of SeqAddressList and DB ;-)

paweljakubas commented on lib/core/test/unit/Cardano/WalletSpec.hs:
```diff
@@ -1330,9 +1331,19 @@ newtype DummyState
-instance Sqlite.PersistState DummyState where
-    insertState _ _ = error "DummyState.insertState: not implemented"
-    selectState _ = error "DummyState.selectState: not implemented"
+instance Sqlite.AddressBookIso DummyState where
```
At some point, at other PR,  maybe we could have more useful DummyState instances here and get additional confidence by adding tests? Do you see something potentially useful here?

HeinrichApfelmus replied:
As far as I understand, the purpose of the `WalletSpec` module is to test that the different layers work well together; however, the individual layers such as the DB layer are tested in other modules. So, I wouldn't make `DummyState` more fancy here, but rather ensure that we have sufficient testing of the individual layers.

In this case, I was wondering whether it was worth testing that the instances of `AddressBookIso` are indeed isomorphisms, but then decided against it, as the will-test-that code is too close to the to-be-tested code, and the current `SqliteSpec` state machine tests would highlight any issues here as well.

paweljakubas replied:
ok, fair enough

Commit:
Add `ErrBadFormat` exception

To be thrown when the database file is in a bad format (missing tables, malformatted entries, other forms of corruption, …)

paweljakubas reviewed (approved):
great PR! very fond of introducing of decoupling between nonchanging and changing parts of the state. And treat them explicitely as separate.
